IronFoundry.Container.Shared/Utilities/EffectiveAccessComputer.cs

```diff
@@ -0,0 +1,116 @@
+                {
+                    throw new Win32Exception(Marshal.GetLastWin32Error());
+                }
+                disposables.Add(authzRM);
+
+
+                // Create the context for the user
+                var securityIdentifier = (SecurityIdentifier) identity.Translate(typeof (SecurityIdentifier));
+                var rawSid = new byte[securityIdentifier.BinaryLength];
+                securityIdentifier.GetBinaryForm(rawSid, 0);
+
+                SafeAuthzContextHandle userClientCtxt;
+                if (!NativeMethods.AuthzInitializeContextFromSid(
+                    NativeMethods.AuthzInitFlags.Default,
+                    rawSid,
+                    authzRM,
+                    IntPtr.Zero,
+                    NativeMethods.LUID.NullLuid,
+                    IntPtr.Zero,
+                    out userClientCtxt))
+                {
+                    throw new Win32Exception(Marshal.GetLastWin32Error());
+                }
+                disposables.Add(userClientCtxt);
+
+
+                // Prepare the Access Check request
+                var request = new NativeMethods.AUTHZ_ACCESS_REQUEST();
+                request.DesiredAccess = ACCESS_MASK.MAXIMUM_ALLOWED;
+                request.PrincipalSelfSid = null;
+                request.ObjectTypeList = IntPtr.Zero;
+                request.ObjectTypeListLength = 0;
+                request.OptionalArguments = IntPtr.Zero;
+
+                // Prepare the access check reply
+                var grantedAccessBuffer = new SafeHGlobal(sizeof (ACCESS_MASK));
+                disposables.Add(grantedAccessBuffer);
+
+                var errorBuffer = new SafeHGlobal(sizeof (uint));
+                disposables.Add(errorBuffer);
+
+                var reply = new NativeMethods.AUTHZ_ACCESS_REPLY();
+                reply.ResultListLength = 1;
+                reply.SaclEvaluationResults = IntPtr.Zero;
+                reply.GrantedAccessMask = grantedAccessBuffer.DangerousGetHandle();
+                reply.Error = errorBuffer.DangerousGetHandle();
+
+                // Do the access check
+                var rawSD = new byte[descriptor.BinaryLength];
+                descriptor.GetBinaryForm(rawSD, 0);
+
+                if (!NativeMethods.AuthzAccessCheck(
+                    NativeMethods.AuthzACFlags.None,
+                    userClientCtxt,
+                    ref request,
+                    IntPtr.Zero,
+                    rawSD,
+                    null,
+                    0,
+                    ref reply,
+                    IntPtr.Zero))
+                {
+                    throw new Win32Exception(Marshal.GetLastWin32Error());
+                }
+
+                accessGranted = (ACCESS_MASK) Marshal.ReadInt32(grantedAccessBuffer.DangerousGetHandle());
+            }
+            finally
+            {
+                // Clean up all the unmanaged memory
+                foreach (var disposable in disposables)
```

Comment:
I would prefer that we use `using` blocks to manage object lifetime vs. a list of disposables.  The list makes more sense if we don't know the exact set of objects that need to be disposed, or if we were only disposing in case of a failure.